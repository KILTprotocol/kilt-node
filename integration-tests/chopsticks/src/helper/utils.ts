import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
