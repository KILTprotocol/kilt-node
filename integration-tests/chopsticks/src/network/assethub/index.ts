import * as Main from './main_network.js'
import * as Test from './test_network.js'

export const chains = {
	main: {
		storage: Main.storage,
		getSetupOptions: Main.getSetupOptions,
		parachainInfo: Main.parachainInfo,
	},
	test: {
		storage: Test.storage,
		getSetupOptions: Test.getSetupOptions,
		parachainInfo: Test.parachainInfo,
	},
}
