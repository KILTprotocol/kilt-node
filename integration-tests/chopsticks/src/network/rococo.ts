import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { initialBalanceROC, toNumber } from '../utils.js'

/// Options used to create the HydraDx context
export const options: SetupOption = {
	endpoint: process.env.ROCOCO_WS || ['wss://rococo-rpc.polkadot.io'],
	db: './db/rococo.db.sqlite',
	port: toNumber(process.env.ROCOCO_PORT) || 8999,
}

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceROC) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

export function setSudoKey(sudo: string) {
	return {
		Sudo: {
			key: sudo,
		},
	}
}

export function removeDisputesAndMessageQueues() {
	return {
		ParasDisputes: {
			// those can makes block building super slow
			$removePrefix: ['disputes'],
		},
		Dmp: {
			// clear existing dmp to avoid impact test result
			$removePrefix: ['downwardMessageQueues'],
		},
	}
}

export async function getContext(): Promise<Config> {
	return setupContext(options)
}
