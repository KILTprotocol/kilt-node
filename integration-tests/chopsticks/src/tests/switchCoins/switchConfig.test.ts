import { test } from 'vitest'

import * as PeregrineConfig from '../../network/peregrine.js'
import * as AssetHubConfig from '../../network/assethub.js'
import * as BasiliskConfig from '../../network/basilisk.js'
import * as RococoConfig from '../../network/rococo.js'
import { initialBalanceKILT, initialBalanceROC, keysAlice, keysBob, keysCharlie } from '../../utils.js'
import {
	peregrineContext,
	getFreeBalancePeregrine,
	getFreeRocPeregrine,
	getFreeEkiltAssetHub,
	assethubContext,
	getFreeRocAssetHub,
	basiliskContext,
	rococoContext,
} from '../index.js'
import { checkBalance, createBlock, setStorage, hexAddress } from '../utils.js'
import { getAccountLocationV3, getChildLocation, getNativeAssetIdLocation } from '../../network/utils.js'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'

test.skip('Switch PILTS against EPILTS not same user', async ({ expect }) => {
	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address], initialBalanceROC),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	await setStorage(peregrineContext, PeregrineConfig.setSwitchPair())

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccounts(
			[keysAlice.address, PeregrineConfig.siblingSovereignAccount],
			initialBalanceROC
		),
		...AssetHubConfig.createForeignAsset(keysCharlie.address, [PeregrineConfig.siblingSovereignAccount]),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))

	// 50 PILTS
	const balanceToTransfer = BigInt('50000000000000000')

	const beneficiary = getAccountLocationV3(hexAddress(keysBob.address))

	let section: string = ''
	let errorName: string = ''

	await peregrineContext.api.tx.assetSwitchPool1
		.switch(balanceToTransfer.toString(), beneficiary)
		.signAndSend(keysAlice, ({ dispatchError }) => {
			if (dispatchError) {
				const decoded = peregrineContext.api.registry.findMetaError(dispatchError.asModule)
				section = decoded.section
				errorName = decoded.name
			}
		})

	// After creating a new block, the tx should be finalized
	await createBlock(peregrineContext)

	expect(section).toBe('assetSwitchPool1')
	expect(errorName).toBe('Hook')

	// Check sender state
}, 20_000)

test.skip('Switch PILTS against EPILTS user has not enough balance', async ({ expect }) => {
	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address], initialBalanceROC),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	await setStorage(peregrineContext, PeregrineConfig.setSwitchPair())

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccounts(
			[keysAlice.address, PeregrineConfig.siblingSovereignAccount],
			initialBalanceROC
		),
		...AssetHubConfig.createForeignAsset(keysCharlie.address, [PeregrineConfig.siblingSovereignAccount]),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))

	// 500 PILTS
	const balanceToTransfer = BigInt('500000000000000000')

	const beneficiary = getAccountLocationV3(hexAddress(keysAlice.address))

	let section: string = ''
	let errorName: string = ''

	await peregrineContext.api.tx.assetSwitchPool1
		.switch(balanceToTransfer.toString(), beneficiary)
		.signAndSend(keysAlice, ({ dispatchError }) => {
			if (dispatchError) {
				const decoded = peregrineContext.api.registry.findMetaError(dispatchError.asModule)
				section = decoded.section
				errorName = decoded.name
			}
		})

	// After creating a new block, the tx should be finalized
	await createBlock(peregrineContext)

	expect(section).toBe('assetSwitchPool1')
	expect(errorName).toBe('UserSwitchBalance')

	// Check sender state
}, 20_000)

test.skip('Switch PILTS against EPILTS not enough pool account balance', async ({ expect }) => {
	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT * BigInt(1000)),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address], initialBalanceROC),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	// create swtich pair and give pool account less coins
	await setStorage(peregrineContext, PeregrineConfig.setSwitchPair(initialBalanceKILT))

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccounts(
			[keysAlice.address, PeregrineConfig.siblingSovereignAccount],
			initialBalanceROC
		),
		...AssetHubConfig.createForeignAsset(keysCharlie.address, [PeregrineConfig.siblingSovereignAccount]),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT * BigInt(1000))
	await checkBalance(getFreeBalancePeregrine, PeregrineConfig.initialPoolAccountId, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))

	// 200 PILTS
	const balanceToTransfer = initialBalanceKILT * BigInt(2)

	const beneficiary = getAccountLocationV3(hexAddress(keysAlice.address))

	let section: string = ''
	let errorName: string = ''

	await peregrineContext.api.tx.assetSwitchPool1
		.switch(balanceToTransfer.toString(), beneficiary)
		.signAndSend(keysAlice, ({ dispatchError }) => {
			if (dispatchError) {
				const decoded = peregrineContext.api.registry.findMetaError(dispatchError.asModule)
				section = decoded.section
				errorName = decoded.name
				console.log(section, errorName)
			}
		})

	// After creating a new block, the tx should be finalized
	await createBlock(peregrineContext)

	expect(section).toBe('assetSwitchPool1')
	expect(errorName).toBe('Liquidity')
}, 20_000)

test.skip('Switch PILTS against EPILTS user has no DOTs', async ({ expect }) => {
	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, []),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	// create switch pair and give pool account less coins
	await setStorage(peregrineContext, PeregrineConfig.setSwitchPair())

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccounts(
			[keysAlice.address, PeregrineConfig.siblingSovereignAccount],
			initialBalanceROC
		),
		...AssetHubConfig.createForeignAsset(keysCharlie.address, [PeregrineConfig.siblingSovereignAccount]),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(
		getFreeBalancePeregrine,
		PeregrineConfig.initialPoolAccountId,
		expect,
		PeregrineConfig.initialRemoteFeeAssetBalance
	)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, BigInt(0))

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))

	// 50 PILTS
	const balanceToTransfer = initialBalanceKILT / BigInt(2)

	const beneficiary = getAccountLocationV3(hexAddress(keysAlice.address))

	let section: string = ''
	let errorName: string = ''

	await peregrineContext.api.tx.assetSwitchPool1
		.switch(balanceToTransfer.toString(), beneficiary)
		.signAndSend(keysAlice, ({ dispatchError }) => {
			if (dispatchError) {
				const decoded = peregrineContext.api.registry.findMetaError(dispatchError.asModule)
				section = decoded.section
				errorName = decoded.name
				console.log(section, errorName)
			}
		})

	// After creating a new block, the tx should be finalized
	await createBlock(peregrineContext)

	expect(section).toBe('assetSwitchPool1')
	expect(errorName).toBe('UserXcmBalance')
}, 20_000)

test.skip('Switch PILTS against EPILTS no SwitchPair', async ({ expect }) => {
	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address]),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccounts(
			[keysAlice.address, PeregrineConfig.siblingSovereignAccount],
			initialBalanceROC
		),
		...AssetHubConfig.createForeignAsset(keysCharlie.address, [PeregrineConfig.siblingSovereignAccount]),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))

	// 50 PILTS
	const balanceToTransfer = initialBalanceKILT / BigInt(2)

	const beneficiary = getAccountLocationV3(hexAddress(keysAlice.address))

	let section: string = ''
	let errorName: string = ''

	await peregrineContext.api.tx.assetSwitchPool1
		.switch(balanceToTransfer.toString(), beneficiary)
		.signAndSend(keysAlice, ({ dispatchError }) => {
			if (dispatchError) {
				const decoded = peregrineContext.api.registry.findMetaError(dispatchError.asModule)
				section = decoded.section
				errorName = decoded.name
				console.log(section, errorName)
			}
		})

	// After creating a new block, the tx should be finalized
	await createBlock(peregrineContext)

	expect(section).toBe('assetSwitchPool1')
	expect(errorName).toBe('SwitchPairNotFound')
}, 20_000)

test.skip('Switch PILTS against EPILTS no enough DOTs on AH', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// Assign alice some KILT and ROC tokens
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address]),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	await setStorage(peregrineContext, PeregrineConfig.setSwitchPair())

	await setStorage(assethubContext, {
		...AssetHubConfig.assignDotTokensToAccounts([PeregrineConfig.siblingSovereignAccount], initialBalanceROC),
		...AssetHubConfig.createForeignAsset(keysCharlie.address, [PeregrineConfig.siblingSovereignAccount]),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)

	// Alice should have NO eKILT on AH
	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))
	await checkBalance(getFreeRocAssetHub, keysAlice.address, expect, BigInt(0))

	// 50 PILTS
	const balanceToTransfer = initialBalanceKILT / BigInt(2)

	const beneficiary = getAccountLocationV3(hexAddress(keysAlice.address))

	const signedTx = peregrineContext.api.tx.assetSwitchPool1
		.switch(balanceToTransfer.toString(), beneficiary)
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	await createBlock(peregrineContext)

	checkEvents(events, 'xcmpQueue').toMatchSnapshot('sender events xcm queue pallet')
	checkEvents(events, 'assetSwitchPool1').toMatchSnapshot('Switch events assetSwitchPool1 pallet')
	checkEvents(events, { section: 'balances', method: 'Transfer' }).toMatchSnapshot('sender events Balances')

	await createBlock(assethubContext)

	// messageQueue should not successfully execute the msg
	checkSystemEvents(assethubContext, 'messageQueue').toMatchSnapshot('receiver events xcm queue pallet')
	// Refunded fees should be trapped
	checkSystemEvents(assethubContext, 'polkadotXcm').toMatchSnapshot('receiver events polkadotXcm')

	await checkBalance(getFreeEkiltAssetHub, keysAlice.address, expect, BigInt(0))
	await checkBalance(getFreeRocAssetHub, keysAlice.address, expect, BigInt(0))

	//await assethubContext.pause()
}, 20_00000)

// Is failing: Todo: Fix XCM config
test.skip('Send DOTs from Relay 2 Peregrine', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// Assign alice some KILTs
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, []),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	await setStorage(peregrineContext, PeregrineConfig.setSwitchPair())

	// Assigned Alice some ROCs and HDX on Basilisk
	await setStorage(rococoContext, RococoConfig.assignNativeTokensToAccounts([keysAlice.address]))

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, BigInt(0))

	// 50 ROCs
	const balanceToTransfer = initialBalanceROC / BigInt(2)

	const aliceAddress = hexAddress(keysAlice.address)
	const hydraDxDestination = { V3: getChildLocation(PeregrineConfig.paraId) }
	const beneficiary = getAccountLocationV3(aliceAddress)
	const assetToTransfer = { V3: [getNativeAssetIdLocation(balanceToTransfer)] }

	const signedTx = rococoContext.api.tx.xcmPallet
		.limitedReserveTransferAssets(hydraDxDestination, beneficiary, assetToTransfer, 0, 'Unlimited')
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	await createBlock(rococoContext)

	// MSG should be successfully send
	checkEvents(events, 'xcmPallet').toMatchSnapshot('sender events xcmPallet')

	await createBlock(peregrineContext)

	// messageQueue should not successfully execute the msg
	checkSystemEvents(peregrineContext, {
		section: 'parachainSystem',
		method: 'DownwardMessagesReceived',
	}).toMatchSnapshot('receiver events parachainSystem pallet')

	checkSystemEvents(peregrineContext, {
		section: 'dmpQueue',
		method: 'ExecutedDownward',
	}).toMatchSnapshot('receiver events dmpQueue pallet')

	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, BigInt(0))
}, 20_000)

// Is failing: Todo: Fix XCM config
test.skip('Send DOTs from basilisk 2 Peregrine', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// Assign alice some KILTs
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, []),
		...PeregrineConfig.setSafeXcmVersion3(),
	})

	await setStorage(peregrineContext, PeregrineConfig.setSwitchPair())

	// Assigned Alice some ROCs and HDX on Basilisk
	await setStorage(basiliskContext, {
		...BasiliskConfig.assignNativeTokensToAccounts([keysAlice.address]),
		...BasiliskConfig.assignRocTokensToAccounts([keysAlice.address], initialBalanceROC),
	})

	// check initial balance of Alice on Spiritnet
	await checkBalance(getFreeBalancePeregrine, keysAlice.address, expect, initialBalanceKILT)
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, BigInt(0))

	// 50 ROCs
	const balanceToTransfer = initialBalanceROC / BigInt(2)

	const beneficiary = {
		V3: {
			parents: 1,
			interior: {
				X2: [
					{ Parachain: PeregrineConfig.paraId },
					{
						AccountId32: {
							id: hexAddress(keysAlice.address),
						},
					},
				],
			},
		},
	}

	const signedTx = basiliskContext.api.tx.xTokens
		.transfer(BasiliskConfig.dotTOkenId, balanceToTransfer, beneficiary, 'Unlimited')
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	await createBlock(basiliskContext)

	// MSG should be successfully send
	checkEvents(events, 'xTokens').toMatchSnapshot('sender events xTokens')
	// tokens have to be withdrawn
	checkEvents(events, 'tokens').toMatchSnapshot('sender events tokens')
	// An upward message should have been sent
	checkEvents(events, 'parachainSystem').toMatchSnapshot('sender events tokens')

	await createBlock(rococoContext)

	// messageQueue should not successfully execute the msg
	checkSystemEvents(rococoContext, 'messageQueue').toMatchSnapshot('relayer events messageQueue')
	// Balance should be moved to peregrine sovereign account
	checkSystemEvents(rococoContext, { section: 'balances', method: 'Minted' }).toMatchSnapshot(
		'relayer events balances minted'
	)
	// Balance should be burned from the basilisk sovereign account
	checkSystemEvents(rococoContext, { section: 'balances', method: 'Burned' }).toMatchSnapshot(
		'relayer events balances Burned'
	)

	await createBlock(peregrineContext)

	// We should still receive the message
	checkSystemEvents(peregrineContext, {
		section: 'parachainSystem',
		method: 'DownwardMessagesReceived',
	}).toMatchSnapshot('receiver events parachainSystem pallet')

	// ... But msg execution should fail
	checkSystemEvents(peregrineContext, {
		section: 'dmpQueue',
		method: 'ExecutedDownward',
	}).toMatchSnapshot('receiver events dmpQueue')

	// // Alice should still have zero balance
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, BigInt(0))
}, 20_000)

test.skip('User gets dusted with ROCs', async ({ expect }) => {
	const { checkEvents } = withExpect(expect)

	// Assign alice some KILTs and ROCs
	await setStorage(peregrineContext, {
		...PeregrineConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...PeregrineConfig.createAndAssignRocs(keysCharlie.address, [keysAlice.address]),
	})

	// 99.9998% of the balance
	const balanceToTransfer = (initialBalanceKILT * BigInt(999998)) / BigInt(1000000)

	// Send all coins to Bob
	const signedTx = peregrineContext.api.tx.balances
		.transferAllowDeath(keysBob.address, balanceToTransfer)
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	await createBlock(peregrineContext)

	checkEvents(events, { section: 'balances', method: 'Transfer' }).toMatchSnapshot('balances transfer event')
	// User should get dusted by this operation
	checkEvents(events, { section: 'balances', method: 'DustLost' }).toMatchSnapshot('balances transfer event')

	// ... But alice's ROC funds should still exist
	await checkBalance(getFreeRocPeregrine, keysAlice.address, expect, initialBalanceROC)
}, 20_000)
