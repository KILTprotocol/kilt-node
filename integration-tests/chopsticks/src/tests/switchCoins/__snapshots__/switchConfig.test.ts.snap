// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Send DOTs from Relay 2 Peregrine > receiver events dmpQueue pallet 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": 1000000000,
        },
      },
    },
    "method": "ExecutedDownward",
    "section": "dmpQueue",
  },
]
`;

exports[`Send DOTs from Relay 2 Peregrine > receiver events parachainSystem pallet 1`] = `
[
  {
    "data": {
      "count": 1,
    },
    "method": "DownwardMessagesReceived",
    "section": "parachainSystem",
  },
]
`;

exports[`Send DOTs from Relay 2 Peregrine > sender events xcmPallet 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(rounded 6200)",
            "refTime": "(rounded 300000000)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "xcmPallet",
  },
  {
    "data": {
      "fees": [
        {
          "fun": {
            "Fungible": "(rounded 130000000)",
          },
          "id": {
            "interior": "Here",
            "parents": 0,
          },
        },
      ],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Rococo",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "xcmPallet",
  },
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": "(rounded 2100)",
            },
          ],
        },
        "parents": 0,
      },
      "message": [
        {
          "ReserveAssetDeposited": [
            {
              "fun": {
                "Fungible": 50000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 50000000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "DepositAsset": {
            "assets": {
              "Wild": {
                "AllCounted": 1,
              },
            },
            "beneficiary": {
              "interior": {
                "X1": [
                  {
                    "AccountId32": {
                      "id": "(hash)",
                      "network": null,
                    },
                  },
                ],
              },
              "parents": 0,
            },
          },
        },
      ],
      "messageId": "(hash)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Rococo",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "xcmPallet",
  },
]
`;

exports[`Send DOTs from basilisk 2 Peregrine > relayer events balances Burned 1`] = `
[
  {
    "data": {
      "amount": 50000000000000,
      "who": "5Ec4AhNugFBakYoh6kD3mgCkUDhy4vVm5o78KDtG42yrXKwS",
    },
    "method": "Burned",
    "section": "balances",
  },
]
`;

exports[`Send DOTs from basilisk 2 Peregrine > relayer events balances minted 1`] = `
[
  {
    "data": {
      "amount": "(rounded 50000000000000)",
      "who": "5Ec4AhNtskxrg56TpEJ8zweU5h4JVUmGgxDqnoE1grqycu6q",
    },
    "method": "Minted",
    "section": "balances",
  },
  {
    "data": {
      "amount": "(rounded 130000000)",
      "who": "5EYCAe5ijiYfyeZ2JJCGq56LmPyNRAKzpG4QkoQkkQNB5e6Z",
    },
    "method": "Minted",
    "section": "balances",
  },
]
`;

exports[`Send DOTs from basilisk 2 Peregrine > relayer events messageQueue 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": "(rounded 2100)",
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": 7200,
        "refTime": "(rounded 630000000)",
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`Send DOTs from basilisk 2 Peregrine > sender events tokens 1`] = `
[
  {
    "data": {
      "amount": 50000000000000,
      "currencyId": 5,
      "who": "bXmKs2MD6WHcgPEjjDNfgAU7nTzxz3zm9DF6xHajqETdpYBP3",
    },
    "method": "Withdrawn",
    "section": "tokens",
  },
]
`;

exports[`Send DOTs from basilisk 2 Peregrine > sender events tokens 2`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "UpwardMessageSent",
    "section": "parachainSystem",
  },
]
`;

exports[`Send DOTs from basilisk 2 Peregrine > sender events xTokens 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 50000000000000,
          },
          "id": {
            "interior": "Here",
            "parents": 1,
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2100)",
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 50000000000000,
        },
        "id": {
          "interior": "Here",
          "parents": 1,
        },
      },
      "sender": "bXmKs2MD6WHcgPEjjDNfgAU7nTzxz3zm9DF6xHajqETdpYBP3",
    },
    "method": "TransferredAssets",
    "section": "xTokens",
  },
]
`;

exports[`Switch PILTS against EPILTS no enough DOTs on AH > Switch events assetSwitchPool1 pallet 1`] = `
[
  {
    "data": {
      "amount": 50000000000000000,
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": {
        "V3": {
          "interior": {
            "X1": {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          },
          "parents": 0,
        },
      },
    },
    "method": "LocalToRemoteSwitchExecuted",
    "section": "assetSwitchPool1",
  },
]
`;

exports[`Switch PILTS against EPILTS no enough DOTs on AH > receiver events polkadotXcm 1`] = `[]`;

exports[`Switch PILTS against EPILTS no enough DOTs on AH > receiver events xcm queue pallet 1`] = `[]`;

exports[`Switch PILTS against EPILTS no enough DOTs on AH > sender events Balances 1`] = `
[
  {
    "data": {
      "amount": 50000000000000000,
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": "4nv4phaKc4EcwENdRERuMF79ZSSB5xvnAk3zNySSbVbXhSwS",
    },
    "method": "Transfer",
    "section": "balances",
  },
]
`;

exports[`Switch PILTS against EPILTS no enough DOTs on AH > sender events xcm queue pallet 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;
